README

		              SOSlib
                   The SBML ODE SOLVER LIBRARY

       For more information about the SBML ODE Solver Library,
       
                               see:
           http://www.tbi.univie.ac.at/~raim/odeSolver/
         		or send email to:
		      odes@tbi.univie.ac.at 
		     


INTRODUCTION

The SBML ODE Solver Library (SOSlib) is a simple command-line tool and
programming library for (1) constructing a system of ordinary differential
equations (ODE) from chemical reaction networks and (2) numerically
integrating the time course of concentrations of chemical species and (3)
basic visualization of model structure and integration results. It is based
on SBML, the XML standard for description of biological reaction networks, 
the SBML library libSBML for parsing SBML and constructing the ODE system,
and on CVODE for numerical integration of the derived system of
ODEs. Optional data visualization modules allow printing of integration
results directly to Grace and drawing graphs of the reaction network, and a
Jacobian interaction graph of the ODE system via graphviz' graph drawing
library.

The SBML ODE Solver can be used as a programming library, similar in style
to the use of libSBML. Please, see the examples directory for usage within 
other programs, and linking issues.

INSTALLATION 
 
The file INSTALL in the main directory gives instructions how to 
build SOSlib with GNU Tools autoconf/automake.

The file README.txt in folder Win32 gives instructions how to
build SOSlib as a Visual C++ project.

More detailed information can be accessed online at
http://www.tbi.univie.ac.at/~raim/odeSolver/doc/

USAGE

After installing the program, type

odeSolver <sbmlfile.xml> [options]

Please start the program with option -h/--help (type `odeSolver -h')
for more information on program and output control via command-line options.

Short tutorials on usage of this and other SOSlib applications can be
found online at 
http://www.tbi.univie.ac.at/~raim/odeSolver/doc/app.html

ARCHITECTURE

Currently, the SOSlib has been tested on Fedora Linux, 
MasOS X/Darwin and Windows. It is written in pure ISO C and
should thus also run on other systems. 
We would highly appreciate any reports! 

BUG REPORTS
Please see online at 
http://www.tbi.univie.ac.at/~raim/odeSolver/support/

BASIC FUNCTIONALITY

Please start the program with option -h/--help for more information on
program and output control via command-line options.

The SOSlib takes a SBML file of level 1 or level 2 (see
http://sbml.org/documents/) as an input. A SBML file describes networks of 
chemical reactions and initial amounts/concentrations of the participating 
chemicals; additionally it can include predefined events and rules. 
The chemical reactions can have a mathemitcal definition of its rate law, 
which is represented by a `kinetic law' in SBML. The `kinetic law' differs 
from a usual rate law, in that it describes change of amount instead of 
concentrations.

The SOSlib then constructs a set of ordinary differential equations (ODEs)
for the rate of change of all chemical species' concentrations (d[x]/dt =
f(x)) from the reactions where the species appear as reactants or
products. ODEs will not be constructed for species concentrations that are
already defined by SBML assignment or rate rules.  A Jacobian matrix of the
resulting ODE system, i.e. the derivatives of the rates towards all other
concentrations (that are defined by an ODE), is constructed, if all
equations are differentiable (i.e. continuous).

The ODE system is constructed as another SBML model, which can also be
printed out when option -o or --printsbml is given.
This SBML can however ignore SBML intentions: the resulting `species' 
can also describe parameters or compartments, and not only chemcial 
species. Every ODE is represented as an SBML `rate rule'. 
Such a pure ODE model, encoded in SBML, could also be directly passed to
the program.

NOTE THAT: the SOSlib can thus also be used as a SOLVER for 
ANY ODE SYSTEM, if the user is willing to ignore SBML definitions and 
describe her ODE system in (potentially wrong!) SBML.


Then the ODE system is passed to a function that calls CVODE to numerically
integrate the ODEs, i.e. calculate the time development of concentrations.
CVODE uses the generated Jacobian matrix or an internal approximation, if one
or more ODEs contain discontinuous expressions, or the user has chosen so.

The results can be printed to the calling terminal (stdout), to a file or 
to XMGRACE, if the optional grace module has been installed.

Additionally the reaction network can be drawn as a bipartite graph, using
the optional module based on GRAPHVIZ graph drawing algorithms and its
graphical output. The dependecies of rates on species concentrations
as given by the Jacobian matrix (if defined), can also be drawn as a graph.
This can sometime be useful to get a quick impression of relevant 
interactions, e.g. feedback loops, in the system.

Please start the program with option -h/--help for more information on
program and output control via command-line options.

Please await the (upcoming) detailed documentation in the `doc' 
directory of one of the next versions of this distribution, 
for more details on the procedure and usage. Contact the authors (email
adresses above), if you have further questions.

Have Fun!


ChangeLog
2010-09-28 Rainer Machne <raim@tbi.univie.ac.at>
	* added lib64 search in config/*.m4 files, patch by Takeshi Abe

2009-02-13 Christoph Flamm <xtof@tbi.univie.ac.at>
	* changed automake/autoconf framework to produce both the static
	  and shared library in a portable manner

2007-09-20 Rainer Machne <raim@tbi.univie.ac.at>
	* implemented thread-save sensitivity - BREAKING SENSE INTERFACES!:
	** moved sensitivity structures from odeModel_t to new odeSense_t
	** separated compilation of odeModel and odeSense functions, therefore
	** removed TCC (doesn't support multiple code states) online compilation
	* added CV_NORMAL_TSTOP mode for CVODES forward solver
	** added CvodeSettings_setTStop to allow saver dynamic value update
	** automatic usage of TSTOP for event and discontinuous models

2007-08-22 Christoph Flamm <xtof@tbi.univie.ac.at>
	* added gcc compilation on-the-fly capability. This solves the
	32/64 bit architecture problems we had with TCC.

2007-06-20 James Lu and Stefan Mueller
	* implemented adjoint solver for discrete observation data
	* etc. ??

2007-06-12 Christoph Flamm <xtof@tbi.univie.ac.at>
	* updated library sources to work with sundials suit v2.3.0

2007-05-11 Rainer Machne <raim@tbi.univie.ac.at>
	* added Tiny C Compiler (TCC) compilation on-the-fly
	capability (problems: TCC only works for 32 bit architecture).

2008-03-10 Rainer Machne <raim@tbi.univie.ac.at>
	* repaired varySettings and batch integration for design points
	as initially intended by Witold Eryk Wolski


src/options.c
/*
  Last changed Time-stamp: <2005-10-12 12:56:15 raim>
  $Id: options.c,v 1.7 2005/10/12 12:52:08 raimc Exp $
*/
#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <string.h>
#include <time.h>


/* own header files */
#include "sbmlsolver/util.h"
#include "sbmlsolver/options.h"

static struct option const long_options[] =
{ {"modelgraph",    no_argument,       0, 'g'},
  {"determinant",   no_argument,       0, 'd'},
  {"printmodel",    no_argument,       0, 'e'},
  {"onthefly",      no_argument,       0, 'f'},
  {"printsbml",     no_argument,       0, 'o'},
  {"matrixgraph",   no_argument,       0, 'm'},
  {"help",          no_argument,       0, 'h'},
  {"interactive",   no_argument,       0, 'i'},
  {"jacobian",      no_argument,       0, 'j'},
  {"jacobianTime",  no_argument,       0, 'y'},
  {"all",           no_argument,       0, 'a'},
  {"rates",         no_argument,       0, 'r'},
  {"reactions",     no_argument,       0, 'k'},
  {"event",         no_argument,       0, 'n'},
  {"steadyState",   no_argument,       0, 's'},
  {"validate",      no_argument,       0, 'v'},
  {"message",       no_argument,       0, 'l'},
  {"write",         no_argument,       0, 'w'},
  {"xmgrace",       no_argument,       0, 'x'},
  {"error",         required_argument, 0,   0},
  {"rerror",        required_argument, 0,   0},
  {"mxstep",        required_argument, 0,   0},
  {"gvformat",      required_argument, 0,   0},  
  {"printstep",     required_argument, 0,   0},
  {"model",         required_argument, 0,   0},
  {"mpath",         required_argument, 0,   0},
  {"param",         required_argument, 0,   0},
  {"schema11",      required_argument, 0,   0},
  {"schema12",      required_argument, 0,   0},
  {"schema21",      required_argument, 0,   0},
  {"spath",         required_argument, 0,   0},  
  {"time",          required_argument, 0,   0},
  
  {NULL, 0, NULL, 0}
};

/* forward declarations of (private) functions */
void initializeOptions (void);
static void processOptions (int argc, char *argv[]);

/**/
void 
decodeCML (int argc, char *argv[])
{
  initializeOptions();
  processOptions(argc, argv);
}

/**/
void
initializeOptions (void)
{
  strcpy(Opt.GvFormat,   "ps");
  strcpy(Opt.ModelPath,  "./");
  strcpy(Opt.Parameter,  "");
  strcpy(Opt.SchemaPath, "./");
  strcpy(Opt.Schema11,   "sbml-l1v1.xsd");
  strcpy(Opt.Schema12,   "sbml-l1v2.xsd");
  strcpy(Opt.Schema21,   "sbml-l2v1.xsd");
  Opt.Error           = 1e-9;
  Opt.RError          = 1e-4;
  Opt.Mxstep          = 10000;
  Opt.PrintStep       = 50;
  Opt.Time            = 1;
  Opt.HaltOnEvent     = 1;
  Opt.InterActive     = 0;
  Opt.Jacobian        = 1;
  Opt.Determinant     = 0;
  Opt.DrawJacobian    = 0;
  Opt.PrintAll        = 0;
  Opt.PrintJacobian   = 0;
  Opt.PrintReactions  = 0;
  Opt.Xmgrace         = 0;
  Opt.DrawReactions   = 0;
  Opt.PrintModel      = 0;
  Opt.PrintODEsToSBML = 0;
  Opt.PrintOnTheFly   = 0;
  Opt.PrintMessage    = 0;
  Opt.Wheel           = 1;
  Opt.SteadyState     = 0;
  Opt.Validate        = 0;
  Opt.Write           = 0;
}


/**/
static void
processOptions (int argc, char *argv[])
{
  int c, option_index = 0;

  /* memorize program name */  
  if (sscanf(argv[0], "%s", Opt.ExeName) == 0) {
    Warn (stderr, "%s:%d processOptions(): No ExeName found",
	  __FILE__, __LINE__);
    usage (EXIT_FAILURE);
  }
  
  /* process command line options */
  while ((c = getopt_long (argc, argv, "adefghijklmnorstvwxy",
                           long_options, &option_index)) != EOF) {
    switch (c) {
    case 0:
      if (strcmp(long_options[option_index].name, "gvformat")==0) {
        char tmp[256];
        if (sscanf(optarg, "%s", tmp) == 0) {
          Warn (stderr, "%s:%d processOptions(): No graph format specified",
                __FILE__, __LINE__);
          usage (EXIT_FAILURE);
        }
        else { strcpy(Opt.GvFormat, tmp); }
      }
      if (strcmp(long_options[option_index].name, "param")==0) {
       char tmp[64];
        if (sscanf(optarg, "%s", tmp) == 0) {
          Warn (stderr, "%s:%d processOptions(): No parameter specified",
                __FILE__, __LINE__);
          usage (EXIT_FAILURE);
        }
        else { strcpy(Opt.Parameter, tmp); }
      }
      if (strcmp(long_options[option_index].name, "error")==0) {
        double tmp;
        if (sscanf(optarg, "%lf", &tmp) == 0) {
          Warn (stderr,
		"%s:%d processOptions(): No abs. Error tolerance specified",
                __FILE__, __LINE__);
          usage (EXIT_FAILURE);
        }
        else { Opt.Error = tmp; }
      }
      if (strcmp(long_options[option_index].name, "rerror")==0) {
        double tmp;
        if (sscanf(optarg, "%lf", &tmp) == 0) {
          Warn (stderr,
		"%s:%d processOptions(): No rel. error tolerance specified",
                __FILE__, __LINE__);
          usage (EXIT_FAILURE);
        }
        else { Opt.RError = tmp; }
      }
      if (strcmp(long_options[option_index].name, "printstep")==0) {
        double tmp;
        if (sscanf(optarg, "%lf", &tmp) == 0) {
          Warn (stderr, "%s:%d processOptions(): No PrintStep specified",
                __FILE__, __LINE__);
          usage (EXIT_FAILURE);
        }
        else { Opt.PrintStep = tmp; }
      }
      if (strcmp(long_options[option_index].name, "mxstep")==0) {
        double tmp;
        if (sscanf(optarg, "%lf", &tmp) == 0) {
          Warn (stderr, "%s:%d processOptions(): No Mxstep specified",
                __FILE__, __LINE__);
          usage (EXIT_FAILURE);
        }
        else { Opt.Mxstep = tmp; }
      }
      if (strcmp(long_options[option_index].name, "model")==0) {
        char tmp[256];
        if (sscanf(optarg, "%s", tmp) == 0) {
          Warn (stderr, "%s:%d processOptions(): No ModelFile specified",
                __FILE__, __LINE__);
          usage (EXIT_FAILURE);
        } else { strcpy(Opt.ModelFile, tmp); }
      }
      if (strcmp(long_options[option_index].name, "mpath")==0) {
        char tmp[256];
        if (sscanf(optarg, "%s", tmp) == 0) {
          Warn (stderr, "%s:%d processOptions(): No ModelPath specified",
                __FILE__, __LINE__);
          usage (EXIT_FAILURE);
        } else { strcpy(Opt.ModelPath, tmp); }
      }      
      if (strcmp(long_options[option_index].name, "schema11")==0) {
        char tmp[256];
        if (sscanf(optarg, "%s", tmp) == 0) {
          Warn (stderr, "%s:%d processOptions(): No Schema L1v1 specified",
                __FILE__, __LINE__);
          usage (EXIT_FAILURE);
        } else { strcpy(Opt.Schema11, tmp); }
      }
      if (strcmp(long_options[option_index].name, "schema12")==0) {
        char tmp[256];
        if (sscanf(optarg, "%s", tmp) == 0) {
          Warn (stderr, "%s:%d processOptions(): No Schema L1v2 specified",
                __FILE__, __LINE__);
          usage (EXIT_FAILURE);
        } else { strcpy(Opt.Schema12, tmp); }
      }
      if (strcmp(long_options[option_index].name, "schema21")==0) {
        char tmp[256];
        if (sscanf(optarg, "%s", tmp) == 0) {
          Warn (stderr, "%s:%d processOptions(): No Schema L2v1 specified",
                __FILE__, __LINE__);
          usage (EXIT_FAILURE);
        } else { strcpy(Opt.Schema21, tmp); }
      }
      if (strcmp(long_options[option_index].name, "spath")==0) {
        char tmp[256];
        if (sscanf(optarg, "%s", tmp) == 0) {
          Warn (stderr, "%s:%d processOptions(): No SchemaPath specified",
                __FILE__, __LINE__);
          usage (EXIT_FAILURE);
        } else { strcpy(Opt.SchemaPath, tmp); }
      }
      if (strcmp(long_options[option_index].name, "time")==0) {
        double tmp;
        if (sscanf(optarg, "%lf", &tmp) == 0) {
          Warn (stderr, "%s:%d processOptions(): No time specified",
                __FILE__, __LINE__);
          usage (EXIT_FAILURE);
        }
        else { Opt.Time = tmp; }
      }
      break;
    case 'a':
      Opt.PrintAll = 1;
      break;
    case 'h':
      usage (EXIT_SUCCESS);
      break;
    case 'i':
      Opt.InterActive = 1;
      break;
    case 'j':
      Opt.Jacobian = 0;
      break;
    case 'y':
      Opt.PrintJacobian = 1;      
      break;
    case 'k':
      Opt.PrintReactions = 1;      
      break;
    case 'x':
      Opt.Xmgrace = 1;
      break;
    case 'g':
      Opt.DrawReactions = 1;
      break;
    case 'm':
      Opt.DrawJacobian = 1;
      break;
    case 'n':
      /* AMF removed 10 June 2005 Opt.PrintOnTheFly = 1; */
      Opt.HaltOnEvent   = 0;
      break;
    case 'e':
      Opt.PrintModel = 1;
      break;
    case 'f':
      Opt.PrintOnTheFly = 1;
      break;
    case 'l':
      Opt.PrintMessage = 1;
      break;      
    case 'o':
      Opt.PrintODEsToSBML = 1;
      break;
    case 'd':
      Opt.Determinant = 1;
      break; 
    case 'r':
      Opt.PrintRates = 1;
      break;
    case 's':
      Opt.SteadyState = 1;
      break;
    case 'v':
      Opt.Validate = 1;
      break;	      
    case 'w':
      Opt.Write = 1;
      break;
    default:
      usage (EXIT_FAILURE);
    }
  }
  /* parse the to file names from the remaining command-line */
  if ( strlen(Opt.ModelFile) > 0 ) return;
  else if ( (optind + 1) <= argc ) strcpy(Opt.ModelFile, argv[optind++]);
  else {
    Warn (stderr, "No ModelFile found on cmd-line");
    usage(EXIT_FAILURE);
  }
}

/**/
void
usage (int status)
{
  fprintf(stderr, "\nUSAGE:  %s <mpath/sbmlfile.xml> [OPTION]\n\n",
	  Opt.ExeName); 
  fprintf(stderr,
    "GENERAL OPTIONS\n"
    " -h, --help            Print (this) usage information.\n"
    " -i, --interactive     Turn on interactive mode\n"
    "     --gvformat <Str>  Set output format for graph drawings (now set \n"
    "                       to: %s); ignored if compiled w/o graphviz)\n"
   "SBML FILE PARSING\n"	  
    " -v, --validate        Validate SBML file before further processing\n"
	  
    "     --model <Str>     SBML file name (not needed!, see USAGE)\n"
    "                       (now set to: %s)\n"
    "     --mpath <Dir>     Set Model File Path\n"
    "                       (now set to: %s)\n"
    "     --schema11 <Str>  Set filename for SBML schema Level 1 Version 1\n"
    "                       (now set to: %s)\n"
    "     --schema12 <Str>  Set filename for SBML schema Level 1 Version 2\n"
    "                       (now set to: %s)\n"
    "     --schema21 <Str>  Set filename for SBML schema Level 2 Version 1\n"
    "                       (now set to: %s)\n"
    "     --spath <Dir>     Set schema file path, absolute or relative to\n"
    "                       model path (now set to: %s)\n"
    "(1) PRINT REACTIONS AND DERIVED ODEs\n"
    " -e, --equations       Print Reactions and derived ODE system\n"
    " -o, --printsbml       Construct ODEs and print as SBML model\n"
    " -g, --modelgraph      Draw bipartite graph of reaction network\n"
    "                       (to .dot text file if compiled w/o graphviz)\n"
    "(2) INTEGRATING\n"
    " -f, --onthefly        Print results during integration\n"
    " -l, --message         Print messages, and integration statistics\n"
    " -j, --jacobian        Toggle use of the jacobian matrix or CVODE's\n"
    "                       internal approximation (default: jacobian)\n"
    " -s, --steadyState     Abort integration at steady states\n"
    " -n, --event           Do not abort on event detection, but keep\n"
    "                       integrating. ACCURACY DEPENDS ON --printstep!!\n"
    "     --param <Str>     Choose a parameter to vary during batch\n"
    "                       integration, from 0 to value in 50 steps\n"
    "     --printstep <Int> Time steps of output, or\n"
    "                       (now set to: %g)\n"
    "     --time <Float>    Integration end time\n"
    "                       (now set to: %g)\n"	  
    "     --error <Float>   Absolute error tolerance during integration\n"
    "                       (now set to: %g)\n"
    "     --rerror <Float>  Relative error tolerance during integration\n"
    "                       (now set to: %g)\n"
    "     --mxstep <Int>    Maximum step number during integration\n"
    "                       (now set to: %g)\n"
    "(3) INTEGRATION RESULTS\n"
    " -a, --all             Print all available results (y/k/r + conc.).\n"
    " -y, --jacobianTime    Print time course of jacobian matrix entries,\n"
    "                       instead of concentrations\n"
    " -k, --reactions       Print time course of the reactions\n"
    "                       (kinetic laws) instead of concentrations\n"
    " -r, --rates           Print time course of the ODEs, instead of\n"
    "                       concentrations\n"	  
    " -w, --write           Write results to file (path/modelfile.xml.dat)\n"  
    " -x, --xmgrace         Print results to XMGrace; uses SBML Names\n"
    "                       instead of Ids (ignored if compiled w/o Grace)\n" 
    " -m, --matrixgraph     Draw species interactions from the jacobian\n"
    "                       matrix at last timepoint of integration\n"
    "                       (to .dot text file if compiled w/o graphviz)\n"
    "\n",
	  Opt.GvFormat,
	  Opt.ModelFile, Opt.ModelPath,
	  Opt.Schema11, Opt.Schema12, Opt.Schema21,  Opt.SchemaPath,
	  Opt.PrintStep, Opt.Time, Opt.Error, Opt.RError, Opt.Mxstep);
  exit (status);
}

/* End of file */
